File: examples/testAI/testai.py
from PyDtUtil import *
from PyDtCore import *
from PyDtAI import *
from PyDtABC import *
from PyDtAnim import *

import sys
import math
import random

class AICharacter:
   
   def __init__(self, scene, camera, waypoint, filename, speed): 
      self.mSpeed = speed
      self.mCharacter = CharacterWrapper("demoMap/SkeletalMeshes/marine.xml")
      self.mCurrentWaypoint = waypoint
      self.mAStar = WaypointAStar()
      self.mScene = scene
      self.mCharacter.SetGroundClamp(scene, 0.0)
      self.SetPosition(self.mCurrentWaypoint)
      scene.AddDrawable(self.mCharacter)
      
      self.mCharacter.AddChild(camera)

   def SetPosition(self, waypoint):
      trans = Transform()
      trans.SetTranslation(waypoint.GetPosition())
      self.mCharacter.SetTransform(trans)      

   def GetCurrentWaypoint(self):
      return self.mCurrentWaypoint
   
   def FindPathAndGoToWaypoint(self, waypoint):
      self.mAStar.Reset(self.mCurrentWaypoint, waypoint)
      result = self.mAStar.FindPath()
      self.mWaypointPath = self.mAStar.GetPath()
      if not self.mWaypointPath:
          return False
      #for waypoint in self.mWaypointPath:
      #   waypoint.SetRenderFlag(Waypoint.RENDER_GREEN)
      for i in range(len(self.mWaypointPath)):
          self.mWaypointPath[i].SetRenderFlag(Waypoint.RENDER_GREEN) 
      self.mWaypointPath[-1].SetRenderFlag(Waypoint.RENDER_RED)
      return True

   def GoToWaypoint(self, dt, waypoint):
      self.mCharacter.RotateToPoint(waypoint.GetPosition(), dt * 3.0)
      self.mCharacter.SetSpeed(-self.mSpeed)
      if not self.mCharacter.IsAnimationPlaying("Walk"):
        self.mCharacter.ClearAllAnimations(0.5)
        self.mCharacter.PlayAnimation("Walk")

   def GetPosition(self):
      trans = Transform()
      self.mCharacter.GetTransform(trans)
      pos = Vec3()
      trans.GetTranslation(pos)
      return pos

   def AmAtWaypoint(self, waypoint):
      pos = self.GetPosition()
      wayPos = waypoint.GetPosition()
      distToX = abs(pos[0] - wayPos[0])
      distToY = abs(pos[1] - wayPos[1])
      return (distToX < 0.1) and ( distToY < 0.1)   

   def ApplyStringPulling(self):
      if len(self.mWaypointPath) >= 2:
         isector = Isector(self.mScene, self.GetPosition(), self.mWaypointPath[1].GetPosition())
         if not isector.Update():
            self.mWaypointPath[0].SetRenderFlag(Waypoint.RENDER_BLUE)
            del self.mWaypointPath[0]
      

   def Update(self, dt):
      if self.mWaypointPath:
         if(self.AmAtWaypoint(self.mWaypointPath[0])):
            self.mCurrentWaypoint = self.mWaypointPath[0]
            del self.mWaypointPath[0]
            self.mCurrentWaypoint.SetRenderFlag(Waypoint.RENDER_BLUE)
         if self.mWaypointPath:
            self.ApplyStringPulling()
            self.GoToWaypoint(dt, self.mWaypointPath[0])
         else:
            self.mCharacter.SetSpeed(0)
      self.mCharacter.Update(dt)
         
class TestAI(Application):

   def Config(self) :
      contextName = GetDeltaRootPath()+"/examples/data/demoMap";
      Project.GetInstance().SetContext(contextName)
      myMap = Project.GetInstance().GetMap("TesttownLt")
      self.LoadMap(myMap)

      WaypointManager.GetInstance().SetDrawWaypoints(1)
      WaypointManager.GetInstance().SetDrawNavMesh(0, 0)
      self.mDrawNavMesh = 0
      
      self.AddDrawable(WaypointManager.GetInstance())    
      
      trans = Transform()

      trans.SetTranslation(-1.0, 5.5, 1.5)
      trans.SetRotation(180.0, -2.0, 0.0)
      self.GetCamera().SetTransform(trans)

      #create overhead camera
      overheadView = View("overhead view")
      overheadView.SetScene( self.GetScene() )
      overheadView.SetKeyboard( self.GetView().GetKeyboard() )
      overheadView.SetMouse( self.GetView().GetMouse() )

      self.AddView( overheadView )
  
      self.mOverheadCamera = Camera()
      self.mOverheadCamera.SetWindow(self.GetWindow())
      self.mOverheadCamera.SetEnabled(0)
      overheadView.SetCamera( self.mOverheadCamera )
      trans.SetTranslation(-1.0, 5.0, 100.0)
      trans.SetRotation(90.0, 270.0, 0.0)
      
      self.mOverheadCamera.SetTransform(trans)
      self.GetCamera().AddChild(self.mOverheadCamera)
      
      self.mWaypointList = WaypointManager.GetInstance().GetWaypoints()
      self.character = AICharacter(self.GetScene(), self.GetCamera(), self.mWaypointList[0], str("marine/marine.rbody"), 10)                 

      self.mWaypoint = self.mWaypointList[1]
      self.character.FindPathAndGoToWaypoint(self.mWaypoint)

   def PreFrame(self, deltaFrameTime) :      
      if self.character.GetCurrentWaypoint().GetID() == self.mWaypoint.GetID():
          foundPathB = False
          while not foundPathB:
             self.mWaypoint = random.choice(self.mWaypointList)
             foundPathB = self.character.FindPathAndGoToWaypoint(self.mWaypoint)
      self.character.Update(deltaFrameTime)
        
   def KeyPressed( self, keyboard, key ) :
      verdict = 0
      if key == int(KeyboardKey.KEY_Space) :            
         self.mOverheadCamera.SetEnabled(not self.mOverheadCamera.GetEnabled())
         self.GetCamera().SetEnabled(not self.GetCamera().GetEnabled())               
         verdict = 1
      elif key == ord('N'):
         WaypointManager.GetInstance().SetDrawNavMesh(not self.mDrawNavMesh, 1)
         self.mDrawNavMesh = not self.mDrawNavMesh
         verdict = 1
      elif key == int(KeyboardKey.KEY_Escape): 
         self.Quit()
         verdict = 1
      elif key == ord('A'): 
         if self.mOverheadCamera.GetEnabled():
            trans = Transform()
            self.mOverheadCamera.GetTransform(trans)
            vec = Vec3()            
            trans.GetTranslation(vec)
            #vec[1] -= (15.5 / 100.0)
            vec[2] -= 1.0
            trans.SetTranslation(vec)
            self.mOverheadCamera.SetTransform(trans)
            verdict = 1
      elif key == ord('Z'): 
         if self.mOverheadCamera.GetEnabled():
            trans = Transform()
            self.mOverheadCamera.GetTransform(trans)
            vec = Vec3()            
            trans.GetTranslation(vec)
            #vec[1] += (15.5 / 100.0)
            vec[2] += 1.0
            trans.SetTranslation(vec)
            self.mOverheadCamera.SetTransform(trans)
            verdict = 1
      return verdict     
                        
SetDataFilePathList (   GetDeltaDataPathList() + ";" +
                        GetDeltaRootPath() + "/examples/data/;" )

app = TestAI('testMap.xml')

app.Config()
app.Run()

File: branches/RB-2.8_to_Trunk_Merge/examples/testAI/testai.cpp
// TestAI.cpp : defines the implementation of the application
#include <dtUtil/mswinmacros.h>
#ifdef DELTA_WIN32
   #pragma warning(push)
   #pragma warning(disable : 4005)
#endif

#include "testai.h"
#include <dtCore/camera.h>
#include <dtCore/transform.h>
#include <dtCore/map.h>
#include <dtCore/project.h>
#include <dtCore/exceptionenum.h>
#include <dtUtil/datapathutils.h>

#include <dtAI/waypoint.h>
#include <dtAI/waypointmanager.h>

#include <dtUtil/mathdefines.h>
#include <dtUtil/log.h>
#include <dtUtil/fileutils.h>
#include <osg/MatrixTransform>

#ifdef DELTA_WIN32
   #pragma warning(pop)
#endif

using namespace dtCore;
using namespace dtABC;
using namespace dtCore;
using namespace dtAI;

////////////////////////////////////////////////////////////////////////////////
TestAI::TestAI(const std::string& pMapFilename, const std::string& configFilename)
   : Application(configFilename)
   , mDrawNavMesh()
   , mMapFilename(pMapFilename)
{
   // Generating a default config file if there isn't one already
   if (!dtUtil::FileUtils::GetInstance().FileExists(configFilename))
   {
      GenerateDefaultConfigFile();
   }
}

////////////////////////////////////////////////////////////////////////////////
TestAI::~TestAI()
{
   // Clean up so that so that dead references won't get called "atexit"
   dtCore::Project::GetInstance().CloseAllMaps(true);
}

////////////////////////////////////////////////////////////////////////////////
void TestAI::Config()
{
   Application::Config();

   // initialize the Waypoint Helper
   WaypointManager::GetInstance();
   GetScene()->AddChild(&WaypointManager::GetInstance());

   // set the waypoint manager to not (redundantly) delete waypoints
   dtAI::WaypointManager::GetInstance().SetDeleteOnClear(false);

   try
   {
      std::string contextName =  dtUtil::GetDeltaRootPath()+"/examples/data";
      Project::GetInstance().SetContext(contextName);
   }
   catch (const dtUtil::Exception& e)
   {
      e.LogException();
   }

   LoadDemoMap(mMapFilename);

   // turn on viewing of waypoints
   WaypointManager::GetInstance().SetDrawWaypoints(true);

   // by default we wont draw the navmesh
   mDrawNavMesh = false;
   WaypointManager::GetInstance().SetDrawNavMesh(mDrawNavMesh, true);

   // set camera offset
   dtCore::Transform trans;
   trans.SetTranslation(-1.0f, 5.5f, 1.5f);
   trans.SetRotation(180.0f, -2.0f, 0.0f);
   GetCamera()->SetTransform(trans);

   // this is needed so OSG knows which camera can generate events
   GetCamera()->GetOSGCamera()->setAllowEventFocus(true);

   // create overhead camera and disable it by default
   dtCore::View *overheadView = new dtCore::View("overhead view");
   overheadView->SetScene( GetScene() );
   AddView( *overheadView );

   mOverheadCamera = new dtCore::Camera();
   mOverheadCamera->SetWindow(GetWindow());
   mOverheadCamera->SetEnabled(false);
   overheadView->SetCamera( mOverheadCamera.get() );

   // this is needed so OSG knows which camera can generate events
   mOverheadCamera->GetOSGCamera()->setAllowEventFocus(false);

   // set overhead camera offset
   trans.SetTranslation(0.0f, -5.0f, 70.0f);
   trans.SetRotation(0.0f, -90.0f, 0.0f);
   trans.Get(mCameraOffset);
   mOverheadCamera->GetMatrixNode()->setMatrix(mCameraOffset);

   // get the first waypoint to spawn the character at
   const WaypointManager::WaypointMap& pContainer = WaypointManager::GetInstance().GetWaypoints();
   if (pContainer.empty())
   {
      LOG_ERROR("Map '" + mMapFilename + "' does not have any valid waypoints");
      exit(1);
   }
   WaypointManager::WaypointMap::const_iterator iter = pContainer.begin();
   const Waypoint* pWaypoint = (*iter).second;

   dtCore::ResourceDescriptor characterFile("SkeletalMeshes:Marine:marine.xml");

//   // spawn our character
   mCharacter = new dtAI::AICharacter(GetScene(), pWaypoint, characterFile, 3);
//
//   // add the two Cameras as children so they get moved along with the character
//   mCharacter->GetCharacter()->AddChild(GetCamera());
//   mCharacter->GetCharacter()->AddChild(mOverheadCamera.get());

   GoToWaypoint(1);

   // seed the random generator
   srand(4);

   CreateHelpLabel();
}

////////////////////////////////////////////////////////////////////////////////
bool TestAI::KeyPressed(const dtCore::Keyboard* keyboard, int key)
{
   switch (key)
   {
   case ' ':
      if (GetCamera()->GetEnabled())
      {
         GetCamera()->SetEnabled(false);
         mOverheadCamera->SetEnabled(true);
         if (mLabel->GetActive())
         {
            GetCamera()->RemoveChild(mLabel.get());
            mOverheadCamera->AddChild(mLabel.get());
         }
      }
      else
      {
         mOverheadCamera->SetEnabled(false);
         GetCamera()->SetEnabled(true);
         if (mLabel->GetActive())
         {
            mOverheadCamera->RemoveChild(mLabel.get());
            GetCamera()->AddChild(mLabel.get());
         }
      }
      return true;

   case 'n':
      mDrawNavMesh = !mDrawNavMesh;
      WaypointManager::GetInstance().SetDrawNavMesh(mDrawNavMesh, true);
      return true;

   case 'a':
      if (mOverheadCamera->GetEnabled())
      {
         mCameraOffset(3,2) -= 1.0f;
         mCameraOffset(3,1) -= 15.5f / 100.0f;
         mOverheadCamera->GetMatrixNode()->setMatrix(mCameraOffset);
      }
      return true;

   case 'z':
      if (mOverheadCamera->GetEnabled())
      {
         mCameraOffset(3,2) += 1.0f;
         mCameraOffset(3,1) += 15.5f / 100.0f;
         mOverheadCamera->GetMatrixNode()->setMatrix(mCameraOffset);
      }
      return true;

   case osgGA::GUIEventAdapter::KEY_Escape:
      Quit();
      return true;

   case osgGA::GUIEventAdapter::KEY_F1:
      mLabel->SetActive(!mLabel->GetActive());
      if (mLabel->GetActive())
      {
         if (GetCamera()->GetEnabled() &&
            GetCamera()->GetChildIndex(mLabel.get()) == GetCamera()->GetNumChildren())
         {
            mOverheadCamera->RemoveChild(mLabel.get());
            GetCamera()->AddChild(mLabel.get());
         }
         else if (mOverheadCamera->GetEnabled() &&
            mOverheadCamera->GetChildIndex(mLabel.get()) == mOverheadCamera->GetNumChildren())
         {
            GetCamera()->RemoveChild(mLabel.get());
            mOverheadCamera->AddChild(mLabel.get());
         }
      }
      return true;

   default:
      break;
   }
   return false;
}

////////////////////////////////////////////////////////////////////////////////
void TestAI::PreFrame(const double deltaFrameTime)
{
   mCharacter->Update(float(deltaFrameTime));

   if (mCharacter->GetCurrentWaypoint() == mCurrentWaypoint)
   {
      // send the character to a random waypoint
      WaypointManager::WaypointMap::size_type pNumWaypoints = WaypointManager::GetInstance().GetWaypoints().size() - 1;
      unsigned pWaypointNum = dtUtil::RandRange(0U, unsigned(pNumWaypoints));
      GoToWaypoint(pWaypointNum);
   }
}

////////////////////////////////////////////////////////////////////////////////
void TestAI::LoadDemoMap(const std::string& pStr)
{
   try
   {
      std::string pContext = Project::GetInstance().GetContext();
      Map& myMap = Project::GetInstance().GetMap(pStr);

      // Since we are in an Application we can simply call...
      LoadMap(myMap);
   }
   catch (const dtUtil::Exception&)
   {
      LOG_ERROR("ERROR: Map Not Found");
   }
}

////////////////////////////////////////////////////////////////////////////////
bool TestAI::GoToWaypoint(int pWaypointNum)
{
   // loop through the waypoints and send our character to the one
   // whose index is pWaypointNum in the WaypointMap contained within WaypointManager
   const WaypointManager::WaypointMap& pWaypoints = WaypointManager::GetInstance().GetWaypoints();
   WaypointManager::WaypointMap::const_iterator iter = pWaypoints.begin();
   WaypointManager::WaypointMap::const_iterator endOfMap = pWaypoints.end();

   bool pHasPath = false;

   for (int i = 0; iter != endOfMap; ++i, ++iter)
   {
      if (i == pWaypointNum)
      {
         pHasPath = mCharacter->FindPathAndGoToWaypoint((*iter).second);
         if (pHasPath)
         {
            mCurrentWaypoint = (*iter).second;
            return true;
         }
         break;
      }
   }
   return false;
}

////////////////////////////////////////////////////////////////////////////////
void TestAI::CreateHelpLabel()
{
   mLabel = new dtABC::LabelActor();
   osg::Vec2 testSize(32.0f, 5.5f);
   mLabel->SetBackSize(testSize);
   mLabel->SetFontSize(0.8f);
   mLabel->SetTextAlignment(dtABC::LabelActor::AlignmentEnum::LEFT_CENTER);
   mLabel->SetText(CreateHelpLabelText());
   mLabel->SetEnableDepthTesting(false);
   mLabel->SetEnableLighting(false);

   GetCamera()->AddChild(mLabel.get());
   dtCore::Transform labelOffset(-17.0f, 50.0f, 10.5f, 0.0f, 90.0f, 0.0f);
   mLabel->SetTransform(labelOffset, dtCore::Transformable::REL_CS);
   AddDrawable(GetCamera());
}

////////////////////////////////////////////////////////////////////////////////
std::string TestAI::CreateHelpLabelText()
{
   std::string testString("");
   testString += "F1: Toggle Help Screen\n";
   testString += "\n";
   testString += "Space: Toggle overhead/first person camera\n";
   testString += "n: Toggle navmesh\n";
   testString += "a: Zoom in overhead camera\n";
   testString += "z: Zoom out overhead camera\n";

   return testString;
}
